File: triad/src/test/java/com/nhaarman/triad/TriadTest.java
/*
 * Copyright 2015 Niek Haarman
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.nhaarman.triad;

import android.app.Activity;
import android.content.Intent;
import android.support.annotation.NonNull;
import java.util.Iterator;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.MockitoAnnotations.initMocks;

public class TriadTest {

    @Mock
    private Triad.Listener<Object> mListener;

    @Mock
    private Screen<Object> mScreen1;

    @Mock
    private Screen<Object> mScreen2;

    @Mock
    private Screen<Object> mScreen3;

    @Before
    public void setUp() {
        initMocks(this);

        doAnswer(new Answer() {
            @Override
            public Object answer(final InvocationOnMock invocationOnMock) throws Throwable {
                Triad.Callback callback = (Triad.Callback) invocationOnMock.getArguments()[2];
                callback.onComplete();
                return null;
            }
        }).when(mListener).forward(any(Screen.class), any(TransitionAnimator.class), any(Triad.Callback.class));

        doAnswer(new Answer() {
            @Override
            public Object answer(final InvocationOnMock invocationOnMock) throws Throwable {
                Triad.Callback callback = (Triad.Callback) invocationOnMock.getArguments()[2];
                callback.onComplete();
                return null;
            }
        }).when(mListener).backward(any(Screen.class), any(TransitionAnimator.class), any(Triad.Callback.class));

        doAnswer(new Answer() {
            @Override
            public Object answer(final InvocationOnMock invocationOnMock) throws Throwable {
                Triad.Callback callback = (Triad.Callback) invocationOnMock.getArguments()[2];
                callback.onComplete();
                return null;
            }
        }).when(mListener).replace(any(Screen.class), any(TransitionAnimator.class), any(Triad.Callback.class));
    }

    @Test(expected = IllegalStateException.class)
    public void startWith_withoutListener_throwsIllegalStateException() {
    /* Given */
        Triad triad = Triad.emptyInstance();

    /* When */
        triad.startWith(mock(Screen.class));
    }

    @Test
    public void startWith_firstTime_forwardsToScreen() {
    /* Given */
        Triad triad = Triad.emptyInstance();
        triad.setListener(mListener);

    /* When */
        triad.startWith(mScreen1);

    /* Then */
        verify(mListener).forward(eq(mScreen1), any(TransitionAnimator.class), any(Triad.Callback.class));
        assertBackstackHasEntries(triad.getBackstack(), mScreen1);
    }

    @Test
    public void startWith_secondTime_doesNotMoveForward() {
    /* Given */
        Triad triad = Triad.emptyInstance();
        triad.setListener(mListener);
        triad.startWith(mScreen1);
        verify(mListener).forward(eq(mScreen1), any(TransitionAnimator.class), any(Triad.Callback.class));

    /* When */
        triad.startWith(mScreen1);

    /* Then */
        verifyNoMoreInteractions(mListener);
        assertBackstackHasEntries(triad.getBackstack(), mScreen1);
    }

    @Test
    public void startWith_withAValidBackstack_doesNotMoveForward() {
    /* Given */
        Triad triad = Triad.newInstance(Backstack.single(mScreen1), mListener);

    /* When */
        triad.startWith(mScreen1);

    /* Then */
        verifyNoMoreInteractions(mListener);
        assertBackstackHasEntries(triad.getBackstack(), mScreen1);
    }

    @Test(expected = IllegalStateException.class)
    public void goTo_withoutHavingABackstack_throwsIllegalStateException() {
    /* Given */
        Triad triad = Triad.emptyInstance();
        triad.setListener(mListener);

    /* When */
        triad.goTo(mScreen1);
    }

    @Test
    public void goTo_withAValidBackstack_forwardsToScreen() {
    /* Given */
        Triad triad = Triad.emptyInstance();
        triad.setListener(mListener);
        triad.startWith(mScreen1);
        verify(mListener).forward(eq(mScreen1), any(TransitionAnimator.class), any(Triad.Callback.class));

    /* When */
        triad.goTo(mScreen2);

    /* Then */
        verify(mListener).forward(eq(mScreen2), any(TransitionAnimator.class), any(Triad.Callback.class));
        assertBackstackHasEntries(triad.getBackstack(), mScreen1, mScreen2);
    }

    @Test(expected = IllegalStateException.class)
    public void popTo_withAnEmptyBackstack_throwsIllegalStateException() {
    /* Given */
        Triad triad = Triad.emptyInstance();
        triad.setListener(mListener);

    /* When */
        triad.popTo(mScreen1);
    }

    @Test
    public void popTo_withScreenNotOnBackstack_forwardsToScreen() {
    /* Given */
        Triad triad = Triad.emptyInstance();
        triad.setListener(mListener);
        triad.startWith(mScreen1);

    /* When */
        triad.popTo(mScreen2);

    /* Then */
        verify(mListener).forward(eq(mScreen2), any(TransitionAnimator.class), any(Triad.Callback.class));
        assertBackstackHasEntries(triad.getBackstack(), mScreen1, mScreen2);
    }

    @Test
    public void popTo_withScreenOnBackstack_backwardsToScreen() {
    /* Given */
        Triad triad = Triad.newInstance(Backstack.of(mScreen1, mScreen2), mListener);

    /* When */
        triad.popTo(mScreen1);

    /* Then */
        verify(mListener).backward(eq(mScreen1), any(TransitionAnimator.class), any(Triad.Callback.class));
        assertBackstackHasEntries(triad.getBackstack(), mScreen1);
    }

    @Test
    public void popTo_withScreenOnTop_doesNothing() {
    /* Given */
        Triad triad = Triad.newInstance(Backstack.single(mScreen1), mListener);

    /* When */
        triad.popTo(mScreen1);

    /* Then */
        verifyNoMoreInteractions(mListener);
        assertBackstackHasEntries(triad.getBackstack(), mScreen1);
    }

    @Test(expected = IllegalStateException.class)
    public void replaceWith_withAnEmptyBackstack_throwsIllegalStateException() {
    /* Given */
        Triad triad = Triad.emptyInstance();
        triad.setListener(mListener);

    /* When */
        triad.replaceWith(mScreen1);
    }

    @Test
    public void replaceWith_withAValidBackstack_replacesScreen() {
    /* Given */
        Triad triad = Triad.newInstance(Backstack.single(mScreen1), mListener);

    /* When */
        triad.replaceWith(mScreen2);

    /* Then */
        verify(mListener).replace(eq(mScreen2), any(TransitionAnimator.class), any(Triad.Callback.class));
        assertBackstackHasEntries(triad.getBackstack(), mScreen2);
    }

    @Test(expected = IllegalStateException.class)
    public void goBack_withAnEmptyBackstack_throwsIllegalStateException() {
    /* Given */
        Triad triad = Triad.emptyInstance();

    /* When */
        triad.goBack();
    }

    @Test
    public void goBack_withSingleBackstackEntry_doesNothing() {
    /* Given */
        Triad triad = Triad.newInstance(Backstack.single(mScreen1), mListener);

    /* When */
        boolean result = triad.goBack();

    /* Then */
        assertThat(result, is(false));
        assertBackstackHasEntries(triad.getBackstack(), mScreen1);
    }

    @Test
    public void goBack_withMoreThanOneBackstackEntry_backwards() {
    /* Given */
        Triad triad = Triad.newInstance(Backstack.of(mScreen1, mScreen2), mListener);

    /* When */
        boolean result = triad.goBack();

    /* Then */
        assertThat(result, is(true));
        assertBackstackHasEntries(triad.getBackstack(), mScreen1);
    }

    @Test(expected = IllegalStateException.class)
    public void forward_withAnEmptyBackstack_throwsIllegalStateException() {
    /* Given */
        Triad triad = Triad.emptyInstance();
        triad.setListener(mListener);

    /* When */
        triad.forward(Backstack.of(mScreen1, mScreen2));
    }

    @Test
    public void forward_withAValidBackstack_forwardsToLastScreen() {
    /* Given */
        Triad triad = Triad.newInstance(Backstack.single(mScreen1), mListener);

    /* When */
        triad.forward(Backstack.of(mScreen2, mScreen3));

    /* Then */
        verify(mListener).forward(eq(mScreen3), any(TransitionAnimator.class), any(Triad.Callback.class));
        assertBackstackHasEntries(triad.getBackstack(), mScreen2, mScreen3);
    }

    @Test(expected = IllegalStateException.class)
    public void backward_withAnEmptyBackstack_throwsIllegalStateException() {
    /* Given */
        Triad triad = Triad.emptyInstance();
        triad.setListener(mListener);

    /* When */
        triad.backward(Backstack.of(mScreen1, mScreen2));
    }

    @Test
    public void backward_withAValidBackstack_backwardsToLastScreen() {
    /* Given */
        Triad triad = Triad.newInstance(Backstack.single(mScreen1), mListener);

    /* When */
        triad.backward(Backstack.of(mScreen2, mScreen3));

    /* Then */
        verify(mListener).backward(eq(mScreen3), any(TransitionAnimator.class), any(Triad.Callback.class));
        assertBackstackHasEntries(triad.getBackstack(), mScreen2, mScreen3);
    }

    @Test(expected = IllegalStateException.class)
    public void replace_withAnEmptyBackstack_throwsIllegalStateException() {
    /* Given */
        Triad triad = Triad.emptyInstance();
        triad.setListener(mListener);

    /* When */
        triad.replace(Backstack.of(mScreen1, mScreen2));
    }

    @Test
    public void replace_withAValidBackstack_replacesToLastScreen() {
    /* Given */
        Triad triad = Triad.newInstance(Backstack.single(mScreen1), mListener);

    /* When */
        triad.replace(Backstack.of(mScreen2, mScreen3));

    /* Then */
        verify(mListener).replace(eq(mScreen3), any(TransitionAnimator.class), any(Triad.Callback.class));
        assertBackstackHasEntries(triad.getBackstack(), mScreen2, mScreen3);
    }

    @Test(expected = IllegalStateException.class)
    public void startActivity_withoutActivityReference_throwsIllegalStateException() {
    /* Given */
        Triad triad = Triad.newInstance(Backstack.single(mScreen1), mListener);

    /* When */
        triad.startActivity(mock(Intent.class));
    }

    @Test
    public void startActivity_withActivityReference_startsActivity() {
    /* Given */
        Activity activity = mock(Activity.class);
        Intent intent = mock(Intent.class);

        Triad triad = Triad.newInstance(Backstack.single(mScreen1), mListener);
        triad.setActivity(activity);

    /* When */
        triad.startActivity(intent);

    /* Then */
        verify(activity).startActivity(intent);
    }

    @Test(expected = IllegalStateException.class)
    public void startActivityForResult_withoutActivityReference_throwsIllegalStateException() {
    /* Given */
        Triad triad = Triad.newInstance(Backstack.single(mScreen1), mListener);

    /* When */
        triad.startActivityForResult(mock(Intent.class), mock(Triad.ActivityResultListener.class));
    }

    @Test
    public void startActivityForResult_withActivityReference_startsActivityForResult() {
    /* Given */
        Activity activity = mock(Activity.class);
        Intent intent = mock(Intent.class);
        Triad.ActivityResultListener resultListener = mock(Triad.ActivityResultListener.class);

        Triad triad = Triad.newInstance(Backstack.single(mScreen1), mListener);
        triad.setActivity(activity);

    /* When */
        triad.startActivityForResult(intent, resultListener);

    /* Then */
        verify(activity).startActivityForResult(eq(intent), anyInt());
    }

    @SuppressWarnings("rawtypes")
    private static void assertBackstackHasEntries(@NonNull final Backstack backstack, @NonNull final Screen... screens) {
        assertThat("Backstack size", backstack.size(), is(screens.length));

        int i = 0;
        for (Iterator<Screen<?>> iterator = backstack.reverseIterator(); iterator.hasNext(); ) {
            Screen screen = iterator.next();
            assertThat(backstack.toString(), screen, is(screens[i]));
            i++;
        }
    }
}

File: triad-core/src/test/java/com/nhaarman/triad/TriadTest.kt
/*
 * Copyright 2016 Niek Haarman
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.nhaarman.triad

import android.app.Activity
import android.content.Intent
import com.nhaarman.expect.expect
import com.nhaarman.mockito.*
import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Before
import org.junit.Test
import org.mockito.InOrder
import org.mockito.Mock
import org.mockito.MockitoAnnotations.initMocks

class TriadTest {

    @Mock
    private lateinit var mListener: Triad.Listener<Any>

    @Mock
    private lateinit var mScreen1: Screen<Any>

    @Mock
    private lateinit var mScreen2: Screen<Any>

    @Mock
    private lateinit var mScreen3: Screen<Any>

    private lateinit var inOrder: InOrder

    @Before
    fun setUp() {
        initMocks(this)
        inOrder = inOrder(mListener)

        doAnswer({ invocationOnMock ->
            val callback = invocationOnMock.arguments[2] as Triad.Callback
            callback.onComplete()
        }).`when`(mListener).forward(any(), any(), any())

        doAnswer({ invocationOnMock ->
            val callback = invocationOnMock.arguments[2] as Triad.Callback
            callback.onComplete()
        }).`when`(mListener).backward(any(), any(), any())

        doAnswer({ invocationOnMock ->
            val callback = invocationOnMock.arguments[2] as Triad.Callback
            callback.onComplete()
        }).`when`(mListener).replace(any(), any(), any())
    }

    @Test(expected = IllegalStateException::class)
    fun startWith_withoutListener_throwsIllegalStateException() {
        /* Given */
        val triad = Triad.emptyInstance()

        /* When */
        triad.startWith(mock())
    }

    @Test
    fun startWith_firstTime_forwardsToScreen() {
        /* Given */
        val triad = Triad.emptyInstance()
        triad.listener = mListener

        /* When */
        triad.startWith(mScreen1)

        /* Then */
        inOrder.verify(mListener).screenPushed(mScreen1)
        inOrder.verify(mListener).forward(eq(mScreen1), any(), any())
        assertBackstackHasEntries(triad.backstack, mScreen1)
    }

    @Test
    fun startWith_secondTime_doesNotMoveForward() {
        /* Given */
        val triad = Triad.emptyInstance()
        triad.listener = mListener
        triad.startWith(mScreen1)
        verify(mListener).screenPushed(mScreen1)
        verify(mListener).forward(eq(mScreen1), any(), any())

        /* When */
        triad.startWith(mScreen1)

        /* Then */
        verifyNoMoreInteractions(mListener)
        assertBackstackHasEntries(triad.backstack, mScreen1)
    }

    @Test
    fun startWith_withAValidBackstack_doesNotMoveForward() {
        /* Given */
        val triad = Triad.newInstance(Backstack.single(mScreen1), mListener)

        /* When */
        triad.startWith(mScreen1)

        /* Then */
        verifyNoMoreInteractions(mListener)
        assertBackstackHasEntries(triad.backstack, mScreen1)
    }

    @Test(expected = IllegalStateException::class)
    fun goTo_withoutHavingABackstack_throwsIllegalStateException() {
        /* Given */
        val triad = Triad.emptyInstance()
        triad.listener = mListener

        /* When */
        triad.goTo(mScreen1)
    }

    @Test
    fun goTo_withAValidBackstack_forwardsToScreen() {
        /* Given */
        val triad = Triad.emptyInstance()
        triad.listener = mListener
        triad.startWith(mScreen1)
        inOrder.verify(mListener).screenPushed(mScreen1)
        inOrder.verify(mListener).forward(eq(mScreen1), any(), any())

        /* When */
        triad.goTo(mScreen2)

        /* Then */
        inOrder.verify(mListener).screenPushed(mScreen2)
        inOrder.verify(mListener).forward(eq(mScreen2), any(), any())
        assertBackstackHasEntries(triad.backstack, mScreen1, mScreen2)
    }

    @Test(expected = IllegalStateException::class)
    fun popTo_withAnEmptyBackstack_throwsIllegalStateException() {
        /* Given */
        val triad = Triad.emptyInstance()
        triad.listener = mListener

        /* When */
        triad.popTo(mScreen1)
    }

    @Test
    fun popTo_withScreenNotOnBackstack_forwardsToScreen() {
        /* Given */
        val triad = Triad.emptyInstance()
        triad.listener = mListener
        triad.startWith(mScreen1)
        inOrder.verify(mListener).screenPushed(mScreen1)
        inOrder.verify(mListener).forward(eq(mScreen1), any(), any())

        /* When */
        triad.popTo(mScreen2)

        /* Then */
        inOrder.verify(mListener).screenPushed(mScreen2)
        inOrder.verify(mListener).forward(eq(mScreen2), any(), any())
        assertBackstackHasEntries(triad.backstack, mScreen1, mScreen2)
    }

    @Test
    fun popTo_withScreenOnBackstack_backwardsToScreen() {
        /* Given */
        val triad = Triad.newInstance(Backstack.of(mScreen1, mScreen2), mListener)

        /* When */
        triad.popTo(mScreen1)

        /* Then */
        inOrder.verify(mListener).screenPopped(mScreen2)
        inOrder.verify(mListener).backward(eq(mScreen1), any(), any())
        assertBackstackHasEntries(triad.backstack, mScreen1)
    }

    @Test
    fun popTo_withMultipleScreensOnBackstack_backwardsToScreen() {
        /* Given */
        val triad = Triad.newInstance(Backstack.of(mScreen1, mScreen2, mScreen3), mListener)

        /* When */
        triad.popTo(mScreen1)

        /* Then */
        inOrder.verify(mListener).screenPopped(mScreen3)
        inOrder.verify(mListener).screenPopped(mScreen2)
        inOrder.verify(mListener).backward(eq(mScreen1), any(), any())
        assertBackstackHasEntries(triad.backstack, mScreen1)
    }

    @Test
    fun popTo_withScreenOnTop_doesNothing() {
        /* Given */
        val triad = Triad.newInstance(Backstack.single(mScreen1), mListener)

        /* When */
        triad.popTo(mScreen1)

        /* Then */
        verifyNoMoreInteractions(mListener)
        assertBackstackHasEntries(triad.backstack, mScreen1)
    }

    @Test(expected = IllegalStateException::class)
    fun replaceWith_withAnEmptyBackstack_throwsIllegalStateException() {
        /* Given */
        val triad = Triad.emptyInstance()
        triad.listener = mListener

        /* When */
        triad.replaceWith(mScreen1)
    }

    @Test
    fun replaceWith_withAValidBackstack_replacesScreen() {
        /* Given */
        val triad = Triad.newInstance(Backstack.single(mScreen1), mListener)

        /* When */
        triad.replaceWith(mScreen2)

        /* Then */
        verify(mListener).screenPopped(mScreen1)
        verify(mListener).screenPushed(mScreen2)
        verify(mListener).replace(eq(mScreen2), any(), any())
        assertBackstackHasEntries(triad.backstack, mScreen2)
    }

    @Test(expected = IllegalStateException::class)
    fun goBack_withAnEmptyBackstack_throwsIllegalStateException() {
        /* Given */
        val triad = Triad.emptyInstance()

        /* When */
        triad.goBack()
    }

    @Test
    fun goBack_withSingleBackstackEntry_doesNotPopScreen() {
        /* Given */
        val triad = Triad.newInstance(Backstack.single(mScreen1), mListener)

        /* When */
        val result = triad.goBack()

        /* Then */
        verify(mListener, never()).screenPopped(mScreen1)
        assertThat(result, `is`(false))
        assertBackstackHasEntries(triad.backstack, mScreen1)
    }

    @Test
    fun goBack_withMoreThanOneBackstackEntry_backwards() {
        /* Given */
        val triad = Triad.newInstance(Backstack.of(mScreen1, mScreen2), mListener)

        /* When */
        val result = triad.goBack()

        /* Then */
        assertThat(result, `is`(true))
        inOrder.verify(mListener).screenPopped(mScreen2)
        inOrder.verify(mListener).backward(eq(mScreen1), any(), any())
        assertBackstackHasEntries(triad.backstack, mScreen1)
    }

    @Test(expected = IllegalStateException::class)
    fun forward_withAnEmptyBackstack_throwsIllegalStateException() {
        /* Given */
        val triad = Triad.emptyInstance()
        triad.listener = mListener

        /* When */
        triad.forward(Backstack.of(mScreen1, mScreen2))
    }

    @Test
    fun forward_withAValidBackstack_forwardsToLastScreen() {
        /* Given */
        val triad = Triad.newInstance(Backstack.of(mScreen1, mScreen2), mListener)

        /* When */
        triad.forward(Backstack.of(mScreen2, mScreen3))

        /* Then */
        inOrder.verify(mListener).screenPopped(mScreen2)
        inOrder.verify(mListener).screenPopped(mScreen1)

        inOrder.verify(mListener).screenPushed(mScreen2)
        inOrder.verify(mListener).screenPushed(mScreen3)

        inOrder.verify(mListener).forward(eq(mScreen3), any(), any())
        assertBackstackHasEntries(triad.backstack, mScreen2, mScreen3)
    }

    @Test(expected = IllegalStateException::class)
    fun backward_withAnEmptyBackstack_throwsIllegalStateException() {
        /* Given */
        val triad = Triad.emptyInstance()
        triad.listener = mListener

        /* When */
        triad.backward(Backstack.of(mScreen1, mScreen2))
    }

    @Test
    fun backward_withAValidBackstack_backwardsToLastScreen() {
        /* Given */
        val triad = Triad.newInstance(Backstack.of(mScreen1, mScreen2), mListener)

        /* When */
        triad.backward(Backstack.of(mScreen2, mScreen3))

        /* Then */

        inOrder.verify(mListener).screenPopped(mScreen2)
        inOrder.verify(mListener).screenPopped(mScreen1)

        inOrder.verify(mListener).screenPushed(mScreen2)
        inOrder.verify(mListener).screenPushed(mScreen3)

        inOrder.verify(mListener).backward(eq(mScreen3), any(), any())
        assertBackstackHasEntries(triad.backstack, mScreen2, mScreen3)
    }

    @Test(expected = IllegalStateException::class)
    fun replace_withAnEmptyBackstack_throwsIllegalStateException() {
        /* Given */
        val triad = Triad.emptyInstance()
        triad.listener = mListener

        /* When */
        triad.replace(Backstack.of(mScreen1, mScreen2))
    }

    @Test
    fun replace_withAValidBackstack_replacesToLastScreen() {
        /* Given */
        val triad = Triad.newInstance(Backstack.of(mScreen1, mScreen2), mListener)

        /* When */
        triad.replace(Backstack.of(mScreen2, mScreen3))

        /* Then */
        inOrder.verify(mListener).screenPopped(mScreen2)
        inOrder.verify(mListener).screenPopped(mScreen1)

        inOrder.verify(mListener).screenPushed(mScreen2)
        inOrder.verify(mListener).screenPushed(mScreen3)

        verify(mListener).replace(eq(mScreen3), any(), any())
        assertBackstackHasEntries(triad.backstack, mScreen2, mScreen3)
    }

    @Test(expected = IllegalStateException::class)
    fun startActivity_withoutActivityReference_throwsIllegalStateException() {
        /* Given */
        val triad = Triad.newInstance(Backstack.single(mScreen1), mListener)

        /* When */
        triad.startActivity(mock())
    }

    @Test
    fun startActivity_withActivityReference_startsActivity() {
        /* Given */
        val activity = mock<Activity>()
        val intent = mock<Intent>()

        val triad = Triad.newInstance(Backstack.single(mScreen1), mListener)
        triad.setActivity(activity)

        /* When */
        triad.startActivity(intent)

        /* Then */
        verify(activity).startActivity(intent)
    }

    @Test(expected = IllegalStateException::class)
    fun startActivityForResult_withoutActivityReference_throwsIllegalStateException() {
        /* Given */
        val triad = Triad.newInstance(Backstack.single(mScreen1), mListener)

        /* When */
        triad.startActivityForResult(mock(), mock())
    }

    @Test
    fun startActivityForResult_withActivityReference_startsActivityForResult() {
        /* Given */
        val activity = mock<Activity>()
        val intent = mock<Intent>()
        val resultListener = mock<Triad.ActivityResultListener>()

        val triad = Triad.newInstance(Backstack.single(mScreen1), mListener)
        triad.setActivity(activity)

        /* When */
        triad.startActivityForResult(intent, resultListener)

        /* Then */
        verify(activity).startActivityForResult(eq(intent), any())
    }

    @SuppressWarnings("rawtypes")
    private fun assertBackstackHasEntries(backstack: Backstack, vararg screens: Screen<Any>) {
        assertThat("Backstack size", backstack.size(), `is`(screens.size))

        var i = 0
        val iterator = backstack.reverseIterator()
        while (iterator.hasNext()) {
            val screen = iterator.next()
            expect(screen).toBe(screens[i]) { backstack.toString() }
            i++
        }
    }
}